File: src/string/LeetCode392.py


class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        
        s_index,t_index = 0,0
        s_len = len(s)
        t_len = len(t)
        
        while (s_index<s_len and t_index<t_len):
            if(s[s_index]== t[t_index]):
                s_index +=1        
            t_index +=1
            
        return s_len == s_index
        

            
if __name__ == "__main__":
    
    res = Solution().isSubsequence('ace', 'abcde')
    print(res)
 

File: src/string/LeetCode392.java
package string;

public class LeetCode392 {
    public static void main(String[] args) {

        BetterSolution solution = new BetterSolution();
        boolean res = solution.isSubsequence("ace", "abcde");
        System.out.println(res);
        
    }

    static class Solution{
        public boolean isSubsequence(String s, String t) {
            int sLen = s.length();
            if( sLen < 1 ) return true;
            
            int tLen = t.length();

            int sIndex = 0;
            for (int i = 0; i < tLen; i++) {
                if (s.charAt(sIndex)== t.charAt(i)){
                    sIndex++;
                    if(sIndex == sLen -1) return true;
                }
            }
            return false;
        }
    }


    static class BetterSolution{
        public boolean isSubsequence(String s, String t) {
            int sLen = s.length();
            int tLen = t.length();

            int sIndex = 0;
            int tIndex = 0;

            while(sIndex<sLen && tIndex<tLen){
                if(s.charAt(sIndex)== t.charAt(tIndex)) sIndex++;

                tIndex++;
            }

            return sIndex == sLen;
        }

    }

}
